polkadot/service/src/lib.rs

```diff
@@ -257,16 +267,62 @@ fn local_testnet_config() -> ChainConfig {
 	])
 }
 
-impl Service {
+struct GenesisBuilder {
+	config: GenesisConfig,
+}
+
+impl client::GenesisBuilder for GenesisBuilder {
+	fn build(self) -> (Header, Vec<(Vec<u8>, Vec<u8>)>) {
+		let storage = self.config.build_externalities();
+		let block = genesis::construct_genesis_block(&storage);
+		(primitives::block::Header::decode(&mut block.header.encode().as_ref()).expect("to_vec() always gives a valid serialisation; qed"), storage.into_iter().collect())
+	}
+}
+
+/// Creates light client and register protocol with the network service
+pub fn new_light(config: Configuration) -> Result<Service<client::light::Backend, client::RemoteCallExecutor<client::light::Backend, CodeExecutor>>, error::Error> { // TODO: light
+	Service::new(move |_, executor, genesis_builder: GenesisBuilder| {
```

Comment:
for an actual service, the logic needs to be fairly different for a light client. for one, we need to introduce a feedback loop between block import and the sync service, where we can fetch things like validator set changes and state proofs before continuing sync.

Reply:
Isn't result of `import_block` is enough? I mean - we can do anything (like check for validators set changes there (in `import_block`)) there && just return error to sync if something goes wrong. So this will affect only small portion inside `Client` implementation && leave the `Service` as is.